### DISCUSSION TITLE
Streamlink ffmpeg options

### DISCUSSION THREAD

**Discussion by Chris230291** [streamlink/streamlink#3749] [Q&A] [answered]
Hello. I was wondering if anyone can tell me if I understand this correctly?
```
streamlink <url> best --ffmpeg-fout mpegts
```
Does that run the stream though ffmpeg and convert the output to mpegts?
What I want is the equivalent of...
```
ffmpeg -re -i <url> -c copy -f mpegts pipe:
```

Any help is apreciated,
Chris

**Comment by bastimeyer** [comment, marked as answer]
https://streamlink.github.io/cli.html#cmdoption-ffmpeg-fout

> `--ffmpeg-fout OUTFORMAT`
> **When muxing streams**, set the output format to OUTFORMAT.

The parameter does only affect streams which need to get remuxed by Streamlink due to the nature of the protocol or options being used, eg. DASH streams, or HLS streams with external audio streams, or streams with external subtitles, but not streams which don't need to get remuxed, eg. regular MPEGTS HLS streams, etc. If you want to change the output format regardless of the input stream, then you need to do it yourself, eg. by piping Streamlink's output to ffmpeg.

```sh
streamlink --stdout ... | ffmpeg -i pipe:0 -c copy -f ...
```

There are several (old) threads on the issue tracker with more information in regards to remuxing and remuxing on-the-fly.

**Reply by Chris230291** [reply]
Thanks for the reply.
Yes I know I can manually pipe it into ffmpeg like in your example.
The issue with that for me is that I am using a python subprocess with list arguments and I'm not sure how to do that command in list form...

```
streamlinkcmd = [
	'streamlink',
	url,
	'best',
	'-O'
]
streamlink_sb = subprocess.Popen(streamlinkcmd, stdin=subprocess.DEVNULL, stderr=subprocess.PIPE, stdout=subprocess.PIPE)
```

It has to be list form because I dont want to use shell commands in my subprocess.

Any ideas?

**Reply by bastimeyer** [reply]
> The issue with that for me is that I am using a python subprocess with list arguments and I'm not sure how to do that command in list form...
> Any ideas?

This is beyond the scope of this discussion board and you'll have to figure this out yourself. Just a hint though, each list item is its own argv item of the process you're launching.

- https://docs.python.org/3/library/subprocess.html?highlight=subprocess#subprocess.Popen
- https://docs.python.org/3/library/shlex.html#shlex.split

But if you're implementing Streamlink into your python script/project, then you should use the Streamlink API directly and not Streamlink's CLI via a subprocess.

**Comment by Chris230291** [comment]
Ok thanks.
I am not sure I understand the original command still... Will it only use ffmpeg to convert to mpegts if the stream isn't already, or if the stream needs to be combined with other data like subtitles?

What I want is to make sure I am always getting mpegts from streamlink.

**Reply by bastimeyer** [reply]
> Will it only use ffmpeg to convert to mpegts if the stream isn't already

As said and stated in the docs, the default muxing format is "matroska" **if** muxing is required (DASH, HLS+audio, subtitles), otherwise there will be no re-muxing at all. `--ffmpeg-fout` allows you to change the format of the muxed output stream. The output format of other stream types depends on the streams themselves. Regular HLS streams are already MPEGTS (as streamlink currently doesn't support HLS mp4 streams), but there are other stream types as well, so if you always want to work with mpegts, then you have to remux it yourself with the example I've shown you earlier.

**Comment by BataBoom** [comment]
This works

/usr/bin/streamlink --player-continuous-http --https-proxy "yourproxy"  --url "hlsvariant://stream.m3u8" --default-stream 540p -O | ffmpeg -re -i pipe:b -vcodec libx264 -s 960x540 -b:v 1800k -c:a aac -b:a 128k -f flv "rtmp://"

if you use "mkfifo b" first in your terminal
